Implement changeset tag and getAllTags. Resolves #612.

joeldenning opened the pull request:
See #612. This is my first pull request to this repository - let me know if I missed any steps for contributing.

joeldenning commented on packages/git/src/index.ts:
```diff
@@ -26,6 +26,16 @@ async function commit(message: string, cwd: string) {
+    const tags = gitCmd.stdout.toString().split("\n");
+    return tags;
+  } else {
+    throw Error("Unable to retrieve git tags");
```
I wasn't sure if this was the best way to report an error or not - note that the stdout and stderr of the git command are being swallowed which could be tricky for people trying to debug.

Andarist replied:
In another PR of mine, I've just been throwing smth like `throw new Error(gitCmd.stderr.toString());`. This seems to address the issue that you have raised with stuff being swallowed. It's probably ain't perfect as that stderr won't be nicely formatted, and the failure code will still be somewhat lost - but it still seems like it provides more information to the user and as such, I would prefer something along those lines.

In the future, we might create a special `GitError` or something to handle this better.

joeldenning replied:
👍🏿  i'll update my code

joeldenning commented on packages/cli/src/commands/tag/index.ts:
```diff
@@ -0,0 +1,19 @@
+export default async function run(cwd: string) {
+  const { packages } = await getPackages(cwd);
+
+  const allExistingTags = await git.getAllTags(cwd);
```
Wasn't sure if I should add a check to verify a clean git status before tagging? As far as I know, git tag will ignore any uncommitted changes, but the user might not have realized they haven't committed everything yet. Thoughts?

joeldenning replied:
As far as I can tell, changesets publish does not verify a clean git status before publishing.

Andarist commented on packages/git/src/index.ts:
```diff
@@ -26,6 +26,16 @@ async function commit(message: string, cwd: string) {
+async function getAllTags(cwd: string): Promise<string[]> {
+  const gitCmd = await spawn("git", ["tag"], { cwd });
+  if (gitCmd.code === 0) {
+    const tags = gitCmd.stdout.toString().split("\n");
```
With my git version/config the `stdout` contains a trailing newline - so to be more precise we'd need this:
```suggestion
    const tags = gitCmd.stdout.toString().trim().split("\n");
```

Andarist replied:
I'm also wondering if this would work on WIndows - is the line separator unified there in this context?

joeldenning replied:
I booted up my windows laptop (despite a bsod!) and tested this out. Everything worked, although I ended up adding a `.filter(Boolean)` because it was extracting an empty string as a tag, even after I added the `.trim()`.

Here's the code I ran on my windows machine as a test:

```js
const spawn = require('spawndamnit');

spawn("git", ["tag"], { cwd: process.cwd() }).then(gitCmd => {
  const tags = gitCmd.stdout.toString().trim().split("\n").filter(Boolean);
  console.log("tags", tags);
})
```

I ran the test code in a repo with no tags, with one tag, and with two tags. It worked in all cases. Regarding windows new line characters, I'm not sure why the `"\n"` works there but I think it's because a lot of Windows programs use `\n\r` in their code, so that code looking for normal new lines or carriage returns all work. It also might be possible that NodeJS's spawn command does some cross-platform stuff when converting the stdout to a node string. Either way, it seems to work well on Windows.

joeldenning replied:
Here's the resulting output:

```sh
chars [
  118, 49, 46, 48, 46, 48,
  10, 118, 49, 46, 48, 46,
  49, 10
]
```

Andarist replied:
That array is an equivalent of `'v1.0.0\nv1.0.1\n'`, with no hidden character of anything like that. So a single `.trim()` should already handle that:
```js
[118, 49, 46, 48, 46, 48, 10, 118, 49, 46, 48, 46, 49, 10 ].map(s => String.fromCharCode(s)).join('').trim().split("\n") // ['v1.0.0', 'v1.0.1']
```

I would prefer to avoid that extra `.filter(Boolean)` unless we can write a test that would fail without it.

Andarist commented on packages/git/src/index.ts:
```diff
@@ -26,6 +26,16 @@ async function commit(message: string, cwd: string) {
 }
 
+async function getAllTags(cwd: string): Promise<string[]> {
+  const gitCmd = await spawn("git", ["tag"], { cwd });
```
I'm slightly worried that tags might not always be available on CI when working with shallow gi clones and similar setups. I'm no git expert so this would require some research on the topic.

joeldenning replied:
I read through https://git-scm.com/docs/git-clone and tested locally with `git clone --depth=1`. I saw that the tags are included in a shallow clone unless the `--no-tags` flag is passed to git clone.

Different CI/CD tools run git clone with a wide variety of flags, so there may be some tools that pass the `--no-tags` flag. I looked at github's actions/checkout and it allows you to configure the `--depth` flag but I see no mention of the `--no-tags` flag. [This documentation](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches) indicates that it's quite easy to get all the git tags and branches, if needed. I didn't look at travis, circle, or any other CI/CD tools.

joeldenning replied:
I think it's reasonable that the `changeset tag` command would require that you've cloned the tags along with the repo. It's in the name :)

Andarist commented on docs/command-line-options.md:
```diff
@@ -139,3 +140,15 @@ pre [exit|enter {tag}]
+
+This is helpful in situations where a different tool, such as `pnpm publish -r`, is used to publish packages instead of changeset. For situations where `changeset publish` is executed, running `changeset tag` is unneeded. See [issue 612](https://github.com/atlassian/changesets/issues/612) for the original inspiration behind this command.
+
+The git tags are created in the format `pkg-name@version-number` and are based on the current version number of the package.json for each package. It is expected that [`changeset version`](#version) is run before `changeset tag`, so that the package.json versions are updated before the git tags are created.
```
> The git tags are created in the format `pkg-name@version-number` and are based on the current version number of the package.json for each package.

This might not be entirely correct. The mentioned format is true for monorepos with defined workspaces, but Changesets also support single-package repositories and to make things more idiomatic in those scenarios the pattern is `v1.0.0`. 

> It is expected that [`changeset version`](#version) is run before `changeset tag`, so that the package.json versions are updated before the git tags are created.

Should we maybe validate this in the `tag` command? We could probably use `@changesets/read` and ensure that there are no pending changesets. WDYT?

joeldenning replied:
Re-reading the code I wrote, I don't see how the `v` will be present in the git tag. Were you saying that you'd like it to do that, to match `changeset publish`? If so, how can I detect whether the code is running in a single-package repo, so that I can insert the `v` in that situation during `changeset tag`?

Andarist replied:
> So I won't plan on updating changesets tag to do that check, to match publish

Good call 👍 

> If so, how can I detect whether the code is running in a single-package repo, so that I can insert the v in that situation during changeset tag?

This is what we currently do when publishing:
https://github.com/atlassian/changesets/blob/c1decd7b162e560b86ded277772fd8c2a89240b1/packages/cli/src/commands/publish/index.ts#L79-L89

Andarist commented on .changeset/nervous-actors-protect.md:
```diff
@@ -0,0 +1,10 @@
+"@changesets/git": minor
+---
+
+Add `changeset tag` and `getAllTags`
+
+The `changeset tag` command creates git tags for each package.
+
+The `getAllTags` function within @changesets/git retrieves all git tags for a repository.
```
I think I would split this changeset file into 2. The `getAllTags` part is not relevant for people reading about the new feature in`@changesets/cli`.

Don't be afraid of describing a single PR from different angles :) Each touched package might often need a different summary of changes (the change should be - IMHO - described primarily from the perspective of that package).

joeldenning replied:
👍🏿  will update

Andarist commented:
@joeldenning no worries. Sorry for my late reply anyway - i was lately busy and didnt have time to work on Changesets much but i think i got some fresh energy for that (still not a lot of time though 😜). Hopefully, i should be more responsive in the coming weeks.

If you dont end up addressing the issues I will eventually pick this up and bring this through a finish line.

I would appreciate an answer on the question asked here though: https://github.com/atlassian/changesets/pull/634#discussion_r725699020 . Im rethinking how to reimplement publishing and this would help me to consider your use case.

joeldenning commented:
I pushed a couple new commits addressing the feedback, and responded to your question above.

Andarist commented on docs/command-line-options.md:
```diff
@@ -139,3 +140,15 @@ pre [exit|enter {tag}]
+
+The tag command creates git tags for the current version of all packages. The tags created are equivalent to those created by [`changeset publish`](#publish), but the `tag` command does not publish anything to npm.
+
+This is helpful in situations where a different tool, such as `pnpm publish -r`, is used to publish packages instead of changeset. For situations where `changeset publish` is executed, running `changeset tag` is unneeded. See [issue 612](https://github.com/atlassian/changesets/issues/612) for the original inspiration behind this command.
```
Both of the things mentioned could be supported by the `changeset publish`. The latter already is, since we literally call `npm publish` in the given directory. My plan for the former is to invoke the appropriate `pack` command and publish that (so replacing would happen within `pnpm`), this would - most likely - require us to also implement running some lifecycle scripts within changeset (instead of relying on them being executed within `npm publish`) but as far as I remember this (running the scripts) is exactly what `pnpm` does under the hood. So it looks like a valid strategy for us too.

Andarist replied:
@zkochan using smth like `pnpm pack && npm publish lib.tar.gz` would work though, right?

zkochan replied:
right

joeldenning commented:
I've responded to all comments and think I've addressed all of the review feedback - let me know if there's anything I looked over or other things to change.

Andarist commented:
@joeldenning thank you for your contribution and for bearing with me. I've made a TODO item for this week to land this - it's very close to that, just some nits and one last look at this. I can take care of all of that myself so we don't end up 🏓 each other any longer.

Andarist commented:
Ufff, managed to land this before the end of the week 😉 

Thank you very much for your contribution ❤️
